src/hidra/receiver/asapo_transfer.py

```diff
@@ -11,29 +13,33 @@
 
 
 class AsapoTransfer:
-    def __init__(self, asapo_worker, signal_host, target_host, target_port, target_dir):
+    def __init__(self, asapo_worker, signal_host, target_host, target_port, target_dir, reconnect_timout):
 
         self.signal_host = signal_host
         self.targets = [[target_host, target_port, 1]]
         self.target_dir = target_dir
         self.asapo_worker = asapo_worker
+        self.reconnect_timout = reconnect_timout
         self.query = Transfer("STREAM_METADATA", self.signal_host)
-        self.query.initiate(self.targets)
 
-    def start_transfer(self):
-        self.query.start()
-        self.run()
+    def start_transfer(self, stop_event):
+        try:
+            self.query.initiate(self.targets)
+            self.query.start()
+            self.run(stop_event)
+        finally:
+            self.stop()
 
-    def run(self):
-        while True:
-            try:
-                [metadata, data] = self.query.get()
-                if metadata is not None and data is not None:
+    def run(self, stop_event):
```

Comment:
`AsapoTransfer` should have its own internal `stop_even` that is set in `AsapoTransfer.stop` (which should not call `self.query.stop`)